sky/backends/backend_utils.py

```diff
@@ -2681,3 +2689,101 @@ def check_stale_runtime_on_remote(returncode: int, stderr: str,
+    try:
+        cloud.check_features_are_supported(
+            launched_resources, {clouds.CloudImplementationFeatures.OPEN_PORTS})
+    except exceptions.NotSupportedError:
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError('Querying endpoints is not supported '
+                             f'for {cloud}.') from None
+
+    config = common_utils.read_yaml(handle.cluster_yaml)
+    port_details = provision_lib.query_ports(repr(cloud),
+                                             handle.cluster_name_on_cloud,
+                                             handle.launched_resources.ports,
+                                             head_ip=handle.head_ip,
+                                             provider_config=config['provider'])
+
+    # Validation before returning the endpoints
+    if endpoint is not None:
+        # If the requested endpoint was not to be exposed
+        port_set = resources_utils.port_ranges_to_set(
+            handle.launched_resources.ports)
+        if endpoint not in port_set:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(f'Port {endpoint} is not exposed '
+                                 'on cluster '
+                                 f'{cluster_record["name"]!r}.')
+        # If the user requested a specific port endpoint, check if it is exposed
+        if endpoint not in port_details:
+            error_msg = (f'Port {endpoint} not exposed yet. '
+                         f'{_ENDPOINTS_RETRY_MESSAGE} ')
+            if handle.launched_resources.cloud.is_same_cloud(
+                    clouds.Kubernetes()):
+                # Add Kubernetes specific debugging info
+                error_msg += (kubernetes_utils.get_endpoint_debug_message())
+            with ux_utils.print_exception_no_traceback():
+                raise RuntimeError(error_msg)
+        return {endpoint: port_details[endpoint][0].url()}
+    else:
+        if not port_details:
+            # If cluster had no ports to be exposed
+            if handle.launched_resources.ports is None:
+                with ux_utils.print_exception_no_traceback():
+                    raise ValueError('Cluster does not have any ports '
+                                     'to be exposed.')
+            # Else ports have not been exposed even though they exist.
+            # In this case, ask the user to retry.
+            else:
+                error_msg = (f'No endpoints exposed yet. '
+                             f'{_ENDPOINTS_RETRY_MESSAGE} ')
+                if handle.launched_resources.cloud.is_same_cloud(
+                        clouds.Kubernetes()):
+                    # Add Kubernetes specific debugging info
+                    error_msg += \
+                        kubernetes_utils.get_endpoint_debug_message()
+                with ux_utils.print_exception_no_traceback():
+                    raise RuntimeError(error_msg)
```

Comment:
I thought about it and we could do it. However, if we keep the port opening [non-blocking](https://github.com/skypilot-org/skypilot/pull/3377#discussion_r1560120112), it might be safer to raise exceptions here instead of returning an empty dict so the caller does not silently get an empty dict if the ports are not open yet. wdyt?

Reply:
I see - I have updated it to return empty dict instead of raising errors in this commit: https://github.com/skypilot-org/skypilot/pull/3377/commits/a0ba0d17f91ad465dbf1bc43e87e2ecde92b5ba4

One side effect of using logging.warning is that it may be hard to suppress in certain cases and the handling now needs to done in the caller (E.g., in cli.py). Perhaps that's ok?

Here's what the errors now looks like:
```
$ sky status --endpoint 8888 k8s2
W 04-29 13:06:03 backend_utils.py:2771] Port 8888 not exposed yet. If the cluster was recently started, please retry after a while. Additionally, make sure your LoadBalancer is configured correctly.
W 04-29 13:06:03 backend_utils.py:2771] To debug, run: kubectl describe service
Aborted!

$ sky status --endpoints k8s2
W 04-29 13:06:04 backend_utils.py:2791] No endpoints exposed yet. If the cluster was recently started, please retry after a while. Additionally, make sure your LoadBalancer is configured correctly.
W 04-29 13:06:04 backend_utils.py:2791] To debug, run: kubectl describe service
Aborted!
```

wdyt?